### PR TITLE
[JENKINS-55330] Use Interceptor to invalid the github tag cache

### PR DESCRIPTION
**Pull request by c9s**

### CODE REVIEW

**Comment by c9s** [comment]
@jglick could you please take a look? thanks

**Comment by jglick** [comment]
Sorry, I do not maintain this plugin and never worked on the tag support.

**Comment by c9s** [comment]
Not quite sure if the interceptor is used. I don't have experience with okhttp and java, would be great if you can point me out the direction. 👍

**Comment by c9s** [comment]
I will find sometime to continue this work.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 1]
```diff
@@ -369,7 +372,33 @@ public static void checkApiUrlValidity(@Nonnull GitHub gitHub, @CheckForNull Sta
...
+                            .build();
+                    }
+                    return originalResponse;
+                }
+            };
+
             OkHttpClient client = new OkHttpClient().setProxy(getProxy(host));
+            // client.interceptors().add(interceptor);
+            client.networkInterceptors().add(interceptor);
+            client.setProxy(getProxy(host));
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: The proxy is already configured three lines up

**Comment by c9s** [comment]
New investigation result:

The upstream package `GHRepository` uses the native java `HttpURLConnection` to connect to GitHub API, therefore network interceptor is not used

```java
HttpURLConnection uc = getConnector().connect(getApiURL("/"));
```

But why the GitHub ref tag API response is cached in the okhttp cache directory?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java (hunk 1 above)
Comment (c9s) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: thanks

Comment (c9s) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: updated

**Comment by c9s** [comment]
OK Just found out that the HttpURLConnection could use a different Connector. 

OkHttp
----
This library comes with a pluggable connector to use different HTTP client implementations
through `HttpConnector`. In particular, this means you can use [OkHttp](http://square.github.io/okhttp/),
so we can make use of it's HTTP response cache.
Making a conditional request against the GitHub API and receiving a 304 response.

**Comment by bitwiseman** [comment]
@c9s 
Thanks for continuing work.  Looks like we need some tests for this change.

**Comment by allir** [comment]
Any news/timeline on this?

**Comment by bitwiseman** [comment]
@c9s @allir 
We need some tests for this change.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 2]
```diff
@@ -370,7 +373,33 @@ public static void checkApiUrlValidity(@Nonnull GitHub gitHub, @CheckForNull Sta
...
             gb.withRateLimitHandler(CUSTOMIZED);
 
+
+            Interceptor interceptor = new Interceptor() {
+                @Override
+                public Response intercept(Chain chain) throws IOException {
+                    Request request = chain.request();
+
+                    // okhttp caches the tag response even if the tag object is not found.
+                    // tags are mutable, therefore we should not cache the references
```
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Wait... no.  
Most information provided by the github api is mutable.  
We can't just turn off caching for all tag queries.  If caching isn't working on tags but is working for other items, this is probably an issue with the `Cache-Control` or `ETag` values provided by the the server or a bug in the Caching implemented in okhttp2, not something that Jenkins is doing wrong.  

There needs to be a lot more data gathered about what is causing this before we take this step.

**Comment by bitwiseman** [comment]
Interceptors don't work with github-api at this time. Closing this.

**Pull request closed without merging**